[[ Bug 15946 ]] Reinstate array building of CGI control value (POST a…

livecodesebastien opened the pull request:
…nd GET)

The arrays can contain subkeys, and taking this into consideration got removed during the refactoring.

livecodesebastien committed:
[[ Bug 15946 ]] Reinstate array building of CGI control value (POST and GET)

The arrays can contain subkeys, and taking this into consideration got removed during the refactoring.

livecodesebastien committed:
[[ Bug 15946 ]] Add more explicit comment about the purpose of cgi_store_control_value

peter-b commented on engine/src/srvcgi.cpp:
```diff
@@ -595,11 +595,125 @@ static void cgi_fetch_valueref_for_key(MCVariable *p_variable, MCNameRef p_key,
+    t_success = p_variable -> setvalueref(t_path . Ptr(), t_path . Size(), false, p_value);
+
+    // Clean keys array
+    t_path . Delete();
```
I thought an `MCAutoArray()` automatically self-deletes? I don't think this (or the `t_success` variable) is necessary...?

livecodesebastien replied:
You're right. Each created `MCNameRef` pushed onto the AutoArray needs to be deleted though (that's what Clean keys array should do).

livecodesebastien committed:
[[ Bug 15946 ]] Fix memory and indexing issue in CGI value parsing

Change the array of value keys to MCAutoNameRefArray
Make sure that the indices are propertly calculated, if the value at the path is a array, but not a sequence

peter-b commented:
@livecode-vulcan review ok 33cffae

livecode-vulcan commented:
💙 review by @peter-b ok 33cffae1a68b260d76cfa1b47d22348ba45f3ce5

livecode-vulcan commented:
😎 test success 33cffae1a68b260d76cfa1b47d22348ba45f3ce5
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/516)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/520)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/515)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/515)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/524)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/518)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/526)
